src/output.cpp

```diff
@@ -698,24 +683,19 @@ std::string JsonOutput::hist_to_str(const std::vector<uint64_t> &values,
       res << ", ";
 
     res << "{";
-    if (i == 0)
-    {
-      res << "\"max\": -1, ";
-    }
-    else if (i == 1)
-    {
-      res << "\"min\": 0, \"max\": 0, ";
-    }
-    else if (i == 2)
+    long low = i - 1, high = i - 1; // covers negative and first 2<<bits slots
+    if (i > (2 << bits))
     {
-      res << "\"min\": 1, \"max\": 1, ";
-    }
-    else
-    {
-      long low = 1ULL << (i - 2);
-      long high = (1ULL << (i - 2 + 1)) - 1;
-      res << "\"min\": " << low << ", \"max\": " << high << ", ";
+      uint32_t power = (i - 1) >> bits, n = 1 << bits, frac = (i - 1) & (n - 1);
+      low = (1ul << (power - 1)) * (n + frac);
+      power = i >> bits;
+      frac = i & (n - 1);
+      high = (1ul << (power - 1)) * (n + frac) - 1;
     }
+    if (low != -1)
+      res << "\"min\": " << low << ", ";
+    res << "\"max\": " << high << ", ";
```

Comment:
Please add some documentation or unit tests, or anything else to help explain this. Or even slower/dumber code. I'd prefer dumber code cuz this is not in the fast path. I will probably never get enough time or be smart enough to understand the math here

Reply:
added detailed explanation in the Text version (and a reference to it), plus made a special case for the negative interval and the ones with only one value, so code is similar to the previous one.

Reply:
@danobi As a broad rule, I'd consider everything that executes during tracing part of the fast path (including text output). It's not uncommon for me to see dropped events because the bpftrace runtime can't keep up with printing events the BPF program is pumping out. If you just meant histograms, then you're probably right that no one will be printing them at such a high frequency. Documentation is always good though!,